allel/stats/roh.py

```diff
@@ -0,0 +1,164 @@
+        to misalignment
+    transition: float
+        Probability of moving between states
+    min_roh: integer
+        Minimum size (bp) to condsider as a ROH. Will depend on contig size
+        and recombination rate. In humans typically > 1Mbp
+    is_accessible: array_like, bool, shape (npositions)
+        Boolean array for each locus in contig describing whether accessible
+        or not.
+    contig_size: int
+        If is_accessible not known/not provided, allows specification of
+        total length of contig.
+
+    Returns
+    -------
+    roh: ndarray, int, shape (n_roh, 2)
+        Span windows of ROH
+    froh: float
+        Proportion of genome in ROH
+
+    """
+
+
+    # this function computes the likely ROH using an HMM model.
+    # define model:
+    ### 2 state model, 3 observable
+    # 0: ROH
+    # 1: normal
+    # 2: inaccessible
+    emission_px = np.array(emission)
+    emission_px.sort()
+
+    # start probability
+    start_prob = np.repeat(1/emission_px.size, emission_px.size)
+
+    # transition between underlying states
+    transitions = _derive_transition_mx(transition, emission_px.size)
+
+    # probability of inaccessible
+    if is_accessible is None:
+        assert contig_size is not None, \
+            "If accessibility not provided must specify size of contig"
+        # may be a mem problem?
+        is_accessible = ChunkedArray(np.repeat(True, contig_size))
+
+    p_accessible = is_accessible.mean()
+    emission_mx = _derive_emission_mx(emission_px, p_accessible)
+
+    # initialize HMM
+    roh_hmm = hmm.MultinomialHMM(n_components=emission_px.size)
+
+    roh_hmm.n_symbols_ = 3
+    roh_hmm.startprob_ = start_prob
+    roh_hmm.transmat_ = transitions
+    roh_hmm.emissionprob_ = emission_mx
+
+    is_heterozygote = g.is_het()
+
+    hz = is_heterozygote[:, ix]
+    pred, obs = _predict_roh_state(roh_hmm, hz, pos, is_accessible)
+    homozygous_windows = _get_state_windows(pred, state=0)
+
+    # filter by roh size
+    roh_sizes = np.diff(homozygous_windows, axis=1).flatten()
+    roh = np.compress(roh_sizes >= min_roh, homozygous_windows, axis=0)
+
+    # return roh and froh
+    return roh, np.diff(roh, axis=1).sum() / is_accessible.size
+
+
+def _predict_roh_state(model, is_het, pos, accessible):
+
+    # assume non-included pos are homozygous reference
+    assert is_het.shape == pos.shape
+
+    # declaration of this object may result in large mem usage
+    observations = ChunkedArray(
+        np.zeros((accessible.size, 1), dtype='int'))
+
+    for i in np.compress(is_het, pos):
+        observations[i - 1] = 1
+
+    for i in np.where(accessible == 0)[0]:
+        observations[i - 1] = 2
+
+    predictions = model.predict(X=observations)
+    return predictions, observations
+
+
+def _derive_emission_mx(prob, pa):
+    # one row per p in prob
+    # hom, het, unobserved
+    mx = [[(1-p) * (1-pa), p * (1-pa), pa] for p in prob]
```

Comment:
Is this right? `pa` is the probability that a base is accessible? If so, should it not be:

```
mx = [[(1-p)*pa, p*pa, 1-pa] for p in prob]
```

...?